syntax = "proto3";

package national_instruments.flex_logger.automation.protocols;

import "ConfigurationBasedSoftware/FlexLogger/Automation/FlexLogger.Automation.Protocols/StartTriggerCondition.proto";
import "ConfigurationBasedSoftware/FlexLogger/Automation/FlexLogger.Automation.Protocols/StopTriggerCondition.proto";
import "ConfigurationBasedSoftware/FlexLogger/Automation/FlexLogger.Automation.Protocols/ValueChangeType.proto";
import "DiagramSdk/Automation/DiagramSdk.Automation.Protocols/Identifiers.proto";
import "google/protobuf/duration.proto";
import "google/protobuf/empty.proto";
import "google/protobuf/timestamp.proto";

// Service interface for a server side logging specification document.
service LoggingSpecificationDocument {
    // RPC call to get the log file base path
    rpc GetLogFileBasePath(GetLogFileBasePathRequest) returns (GetLogFileBasePathResponse) {}
    // RPC call to set the log file base path
    rpc SetLogFileBasePath(SetLogFileBasePathRequest) returns (SetLogFileBasePathResponse) {}
    // RPC call to get the log file name
    rpc GetLogFileName(GetLogFileNameRequest) returns (GetLogFileNameResponse) {}
    // RPC call to set the log file name
    rpc SetLogFileName(SetLogFileNameRequest) returns (SetLogFileNameResponse) {}
    // RPC call to get the description
    rpc GetLogFileDescription(GetLogFileDescriptionRequest) returns (GetLogFileDescriptionResponse) {}
    // RPC call to set the description
    rpc SetLogFileDescription(SetLogFileDescriptionRequest) returns (google.protobuf.Empty) {}
    // RPC call to get items from the data files pane
    rpc GetLogFiles(GetLogFilesRequest) returns (GetLogFilesResponse) {}
    // RPC call to clear the data files pane
    rpc RemoveLogFiles(RemoveLogFilesRequest) returns (google.protobuf.Empty) {}
    // RPC call to get all test properties
    rpc GetTestProperties(GetTestPropertiesRequest) returns (GetTestPropertiesResponse) {}
    // RPC call to set all test properties
    rpc SetTestProperties(SetTestPropertiesRequest) returns (google.protobuf.Empty) {}
    // RPC call to get a specific test property
    rpc GetTestProperty(GetTestPropertyRequest) returns (GetTestPropertyResponse) {}
    // RPC call to set a specific test property
    rpc SetTestProperty(SetTestPropertyRequest) returns (SetTestPropertyResponse) {}
    // RPC call to remove a specific test property
    rpc RemoveTestProperty(RemoveTestPropertyRequest) returns (RemoveTestPropertyResponse) {}
    // RPC call to get the resolved log file base path
    rpc GetResolvedLogFileBasePath(GetResolvedLogFileBasePathRequest) returns (GetResolvedLogFileBasePathResponse) {}
    // RPC call to get the resolved log file name
    rpc GetResolvedLogFileName(GetResolvedLogFileNameRequest) returns (GetResolvedLogFileNameResponse) {}
    // RPC call to get the start trigger settings
    rpc GetStartTriggerSettings(GetStartTriggerSettingsRequest) returns (GetStartTriggerSettingsResponse) {}
    // RPC call to get the stop trigger settings
    rpc GetStopTriggerSettings(GetStopTriggerSettingsRequest) returns (GetStopTriggerSettingsResponse) {}
    // RPC call to set the start trigger settings to Test Start
    rpc SetTestStartTriggerSettings(SetTestStartTriggerSettingsRequest) returns (google.protobuf.Empty) {}
    // RPC call to set the start trigger settings to Channel value change
    rpc SetValueChangeStartTriggerSettings(SetValueChangeStartTriggerSettingsRequest) returns (google.protobuf.Empty) {}
    // RPC call to set the start trigger settings to Absolute time
    rpc SetTimeStartTriggerSettings(SetTimeStartTriggerSettingsRequest) returns (google.protobuf.Empty) {}
    // RPC call to set the stop trigger settings to Test stop
    rpc SetTestStopTriggerSettings(SetTestStopTriggerSettingsRequest) returns (google.protobuf.Empty) {}
    // RPC call to set the stop trigger settings to Channel value change
    rpc SetValueChangeStopTriggerSettings(SetValueChangeStopTriggerSettingsRequest) returns (google.protobuf.Empty) {}
    // RPC call to set the stop trigger settings to Test time elapsed
    rpc SetTimeStopTriggerSettings(SetTimeStopTriggerSettingsRequest) returns (google.protobuf.Empty) {}
    // RPC call to get the retrigerring configuration
    rpc IsRetriggeringEnabled(IsRetriggeringEnabledRequest) returns (IsRetriggeringEnabledResponse) {}
    // RPC call to set the retrigerring configuration
    rpc SetRetriggering(SetRetriggeringRequest) returns (google.protobuf.Empty) {}
}

// Request object for getting the log file base path
message GetLogFileBasePathRequest {
    // The id for the logging specification document
    national_instruments.diagram_sdk.automation.protocols.ElementIdentifier document_identifier = 1;
}

// Response object for a get log file base path request
message GetLogFileBasePathResponse {
    // The log file base path
    string log_file_base_path = 1;
}

// Request object for setting the log file base path
message SetLogFileBasePathRequest {
    // The id for the logging specification document
    national_instruments.diagram_sdk.automation.protocols.ElementIdentifier document_identifier = 1;
    // The log file base path
    string log_file_base_path = 2;
}

// Response object for a set log file base path request
message SetLogFileBasePathResponse {
}

// Request object for getting the log file name
message GetLogFileNameRequest {
    // The id for the logging specification document
    national_instruments.diagram_sdk.automation.protocols.ElementIdentifier document_identifier = 1;
}

// Response object for a get log file name request
message GetLogFileNameResponse {
    // The log file name
    string log_file_name = 1;
}

// Request object for setting the log file name
message SetLogFileNameRequest {
    // The id for the logging specification document
    national_instruments.diagram_sdk.automation.protocols.ElementIdentifier document_identifier = 1;
    // The log file name
    string log_file_name = 2;
}

// Response object for a set log file name
message SetLogFileNameResponse {
}

// Request object for getting the log file description
message GetLogFileDescriptionRequest {
    // The id for the logging specification document
    national_instruments.diagram_sdk.automation.protocols.ElementIdentifier document_identifier = 1;
}

// Response object for a get log file description request
message GetLogFileDescriptionResponse {
    // The log file description
    string log_file_description = 1;
}

// Request object for setting the log file description
message SetLogFileDescriptionRequest {
    // The id for the logging specification document
    national_instruments.diagram_sdk.automation.protocols.ElementIdentifier document_identifier = 1;
    // The log file description
    string log_file_description = 2;
}

// Log file types
enum LogFileType {
    // TDMS files
    TDMS = 0;
    // CSV files
    CSV = 1;
    // TDMS backup files
    TDMS_BACKUP = 2;
}

// Request object for GetLogFiles
message GetLogFilesRequest {
    // The id for the logging specification document
    national_instruments.diagram_sdk.automation.protocols.ElementIdentifier document_identifier = 1;
    // The type of log files to get
    LogFileType log_file_type = 2;
}

// Response object for GetLogFiles
message GetLogFilesResponse {
    // The full paths on disk of the log files, sorted chronologically by file creation time
    repeated string log_files = 1;
}

// Request object for RemoveLogFiles
message RemoveLogFilesRequest {
    // The id for the logging specification document
    national_instruments.diagram_sdk.automation.protocols.ElementIdentifier document_identifier = 1;
    // Delete files on disk?
    bool delete_files = 2;
}

// Message that defines an individual test property
message TestProperty {
    string property_name = 1;
    string property_value = 2;
    bool prompt_on_start = 3;
}

// Request object for getting all test properties
message GetTestPropertiesRequest {
    // The id for the logging specification document
    national_instruments.diagram_sdk.automation.protocols.ElementIdentifier document_identifier = 1;
}

// Response object for getting all test properties
message GetTestPropertiesResponse {
    // The test properties for the logging specification document. Can be empty.
    repeated TestProperty test_properties = 1;
}

// Request object for setting all test properties
message SetTestPropertiesRequest {
    // The id for the logging specification document
    national_instruments.diagram_sdk.automation.protocols.ElementIdentifier document_identifier = 1;
    // The test properties for the logging specification document. Can be empty.
    repeated TestProperty test_properties = 2;
}

// Request object for getting a specific test property
message GetTestPropertyRequest {
    // The id for the logging specification document
    national_instruments.diagram_sdk.automation.protocols.ElementIdentifier document_identifier = 1;
    // The name of the test property to get
    string property_name = 2;
}

// Response object for getting a specific test property
message GetTestPropertyResponse {
    // The requested test property
    TestProperty test_property = 1;
}

// Request object for setting a specific test property
message SetTestPropertyRequest {
    // The id for the logging specification document
    national_instruments.diagram_sdk.automation.protocols.ElementIdentifier document_identifier = 1;
    // The test property to set
    TestProperty test_property = 2;
}

// Response object for setting a specific test property
message SetTestPropertyResponse {
}

// Request object for removing a specific test property
message RemoveTestPropertyRequest {
    // The id for the logging specification document
    national_instruments.diagram_sdk.automation.protocols.ElementIdentifier document_identifier = 1;
    // The name of the test property to remove
    string property_name = 2;
}

// Response object for removing a specific test property
message RemoveTestPropertyResponse {
}

// Request object for getting the resolved log file base path
message GetResolvedLogFileBasePathRequest {
    // The id for the logging specification document
    national_instruments.diagram_sdk.automation.protocols.ElementIdentifier document_identifier = 1;
}

// Response object for a get resolved log file base path request
message GetResolvedLogFileBasePathResponse {
    // The resolved log file base path. The resolved base path has all placeholders converted to actual values.
    string resolved_log_file_base_path = 1;
}

// Request object for getting the resolved log file name
message GetResolvedLogFileNameRequest {
    // The id for the logging specification document
    national_instruments.diagram_sdk.automation.protocols.ElementIdentifier document_identifier = 1;
}

// Response object for a get resolved log file name request
message GetResolvedLogFileNameResponse {
    // The resolved log file name.  The resolved file name has all placeholders converted to actual values.
    string resolved_log_file_name = 1;
}

// Request object for getting the start trigger settings
message GetStartTriggerSettingsRequest {
    // The id for the logging specification document
    national_instruments.diagram_sdk.automation.protocols.ElementIdentifier document_identifier = 1;
}

// Response object for a getting the start trigger settings
message GetStartTriggerSettingsResponse {
    // The start trigger condition
    StartTriggerCondition start_trigger_condition = 1;
    // The start trigger settings
    string start_trigger_settings = 2;
}

// Request object for getting the stop trigger settings
message GetStopTriggerSettingsRequest {
    // The id for the logging specification document
    national_instruments.diagram_sdk.automation.protocols.ElementIdentifier document_identifier = 1;
}

// Response object for getting the stop trigger settings
message GetStopTriggerSettingsResponse {
    // The stop trigger condition
    StopTriggerCondition stop_trigger_condition = 1;
    // The stop trigger settings
    string stop_trigger_settings = 2;
}

// Request object for SetTestStartTriggerSettings
message SetTestStartTriggerSettingsRequest {
    // The id for the logging specification document
    national_instruments.diagram_sdk.automation.protocols.ElementIdentifier document_identifier = 1;
}

// Request object for SetValueChangeStartTriggerSettings
message SetValueChangeStartTriggerSettingsRequest {
    // The id for the logging specification document
    national_instruments.diagram_sdk.automation.protocols.ElementIdentifier document_identifier = 1;
    // The channel name
    string channel_name = 2;
    // The value change condition
    ValueChangeType value_change_type = 3;
    // The threshold
    double threshold = 4;
    // The min value
    double min_value = 5;
    // The max value
    double max_value = 6;
    // The leading time to include in seconds
    double leading_time = 7;
}

// Request object for SetTimeStartTriggerSettings
message SetTimeStartTriggerSettingsRequest {
    // The id for the logging specification document
    national_instruments.diagram_sdk.automation.protocols.ElementIdentifier document_identifier = 1;
    // The time to start the test
    google.protobuf.Timestamp time = 2;
}

// Request object for SetTestStopTriggerSettings
message SetTestStopTriggerSettingsRequest {
    // The id for the logging specification document
    national_instruments.diagram_sdk.automation.protocols.ElementIdentifier document_identifier = 1;
}

// Request object for SetValueChangeStopTriggerSettings
message SetValueChangeStopTriggerSettingsRequest {
    // The id for the logging specification document
    national_instruments.diagram_sdk.automation.protocols.ElementIdentifier document_identifier = 1;
    // The channel name
    string channel_name = 2;
    // The value change condition
    ValueChangeType value_change_type = 3;
    // The threshold
    double threshold = 4;
    // The min value
    double min_value = 5;
    // The max value
    double max_value = 6;
    // The leading time to include in seconds
    double trailing_time = 7;
}

// Request object for SetTimeStopTriggerSettings
message SetTimeStopTriggerSettingsRequest {
    // The id for the logging specification document
    national_instruments.diagram_sdk.automation.protocols.ElementIdentifier document_identifier = 1;
    // The duration
    google.protobuf.Duration duration = 2;
}

// Request object for IsRetriggeringEnabled
message IsRetriggeringEnabledRequest {
    // The id for the logging specification document
    national_instruments.diagram_sdk.automation.protocols.ElementIdentifier document_identifier = 1;
}

// Response object for IsRetriggeringEnabled
message IsRetriggeringEnabledResponse {
    // The retriggering configuration
    bool is_retriggering_enabled = 1;
}

// Request object for SetRetriggering
message SetRetriggeringRequest {
    // The id for the logging specification document
    national_instruments.diagram_sdk.automation.protocols.ElementIdentifier document_identifier = 1;
    // The triggering configuration
    bool is_retriggering_enabled = 2;
}
